src/julia-syntax.scm

```diff
@@ -2591,27 +2645,53 @@
                                           '()
                                           (append (map cons need-rename renamed)
                                                   (map cons need-rename-def renamed-def))
-                                          scope)))
+                                          scope
+                                          (and soft? (null? argnames))
+                                          hard?
+                                          (if toplevel?
+                                              implicit-globals
+                                              (scope:implicit-globals scope))
+                                          warn-vars)
+                              '()
+                              loc))
             (append! (map (lambda (v) `(local ,v)) newnames)
                      (map (lambda (v) `(local-def ,v)) newnames-def)))
            ))
         ((eq? (car e) 'module)
          (error "\"module\" expression not at top level"))
         ((eq? (car e) 'break-block)
          `(break-block ,(cadr e) ;; ignore type symbol of break-block expression
-                       ,(resolve-scopes- (caddr e) scope))) ;; body of break-block expression
+                       ,(resolve-scopes- (caddr e) scope '() loc))) ;; body of break-block expression
         ((eq? (car e) 'with-static-parameters)
          `(with-static-parameters
-           ,(resolve-scopes- (cadr e) scope (cddr e))
+           ,(resolve-scopes- (cadr e) scope (cddr e) loc)
            ,@(cddr e)))
         ((and (eq? (car e) 'method) (length> e 2))
          `(method
            ,(resolve-scopes- (cadr   e) scope)
            ,(resolve-scopes- (caddr  e) scope)
            ,(resolve-scopes- (cadddr e) scope (method-expr-static-parameters e))))
         (else
+         (if (and (eq? (car e) '=) (symbol? (cadr e))
+                  scope (null? (lam:vars (scope:lam scope)))
+                  (warn-var?! (cadr e) scope)
+                  (= *scopewarn-opt* 1))
+             (let* ((v    (cadr e))
+                    (loc  (extract-line-file loc))
+                    (line (if (= (car loc) 0) (julia-current-line) (car loc)))
+                    (file (if (eq? (cadr loc) 'none) (julia-current-file) (cadr loc))))
+               (julia-logmsg
```

Comment:
This file should not be doing IO

Reply:
Well, this is how we did deprecation warnings. I'm sympathetic to not wanting side effects during lowering, but the alternative seems pretty complex. We can't just move it to run time since we don't want the warning more than once. So `eval` would have to search for this within an expression first, plus the rest of the compiler would have to be able to handle the new form. Is that really worth it?

Reply:
Yeah, we said it was a bad idea then, but was only going to be used in v0.7 and future code wouldn't have any warnings so we didn't bother trying to implement it any better. The thinking was that if someone wanted to be adding warnings again, it should be kinda hard 😝. But since the compiler already does that pre-processing step, it shouldn't be that hard (though ideally perhaps, lowering would just put it in a top-most first-statement position—but if that's awkward to implement right now, I don't think it has any actual benefits).